### ISSUE TITLE
Include value of `Gem.source_date_epoch` in metadata of built gems.

### ISSUE THREAD

**Issue by duckinator** [rubygems/rubygems#3118]
We now (via #2882 and #3088) have builds that are reproducible by default _if you know the value of `Gem.source_date_epoch`/`Gem.source_date_epoch_string`_.

If built gems include the value of `Gem.source_date_epoch_string` in their metadata, we open up the possibility of making automated tools that could detect some instances of things like the [backdoored version of the `bootstrap-sass` gem](https://snyk.io/blog/malicious-remote-code-execution-backdoor-discovered-in-the-popular-bootstrap-sass-ruby-gem/) that was released in March 2019.

I'm not sure, however, where this metadata should be included.

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [x] Publishing a library
  - [x] The command line `gem`
  - [ ] Other

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by rubyFeedback** [comment]
This sounds really great - would be epic if we could get the whole of ruby (MRI) and default gem to fulfil reproducibility criteria e. g. debian's reproducible packages project ( https://wiki.debian.org/ReproducibleBuilds#Even_more ).

3.0 release will be on xmas so about ~10 months still left.

**Comment by deivid-rodriguez** [comment]
@duckinator It seems that as long as the `Gem::Specification#date` is not explicitly set by the gem author, we set it to the date corresponding to `Gem.source_date_epoch`, so this should be already available by converting the date in the specification of the built gem back to a timestamp?

**Comment by duckinator** [comment]
@deivid-rodriguez if that's the case, yeah, it should work. And it already being there is fantastic. I'll try to make a script that demonstrates using this to try to reproduce a build based on a gem file relatively soon. (I'd try it now, but it's after midnight and I have things to do tomorrow lol)

**Comment by duckinator** [comment]
<details>
<summary>Rough first attempt at such a script, which doesn't seem to work yet</summary>

```ruby
#!/usr/bin/env ruby

# Rebuild a gem from source using `rake build`, with SOURCE_DATE_EPOCH
# set to a timestamp derived from the value of the `date` field in the gem
# metadata.

require 'date'
require 'digest'
require 'rubygems/package'

DATE_FORMAT = '%Y-%m-%d %H:%M:%S.%N Z'

def print_sha256(file)
  hash = Digest::SHA256.hexdigest(File.read(file))
  puts "#{file}\t#{hash}"
end

def rebuild(original_file, output_file)
  date = Gem::Package.new(original_file).spec.date

  timestamp = date.to_time.to_i

  puts "Original date:      #{date}"
  puts "Reconstructed date: #{date}"
  puts "Reconestructed timestamp: #{timestamp}"

  system({"SOURCE_DATE_EPOCH" => timestamp.to_s}, "rake build")

  print_sha256(original_file)
  print_sha256(output_file)
end

if ARGV.length != 2
  puts "Usage: $0 ORIGINAL_FILE.gem OUTPUT_FILE.gem"
  exit 1
else
  rebuild(ARGV[0], ARGV[1])
end
```

</details>

Still TBD whether _inq_ was a bad choice of gem, the script doesn't work, or what:

```
~/dev/duckinator/inq$ bundle exec rake build
inq 26.0.0 built to pkg/inq-26.0.0.gem.
~/dev/duckinator/inq$ mv pkg/inq-26.0.0.gem{,-old} 
~/dev/duckinator/inq$ bundle exec ../gem-rebuild.rb pkg/inq-26.0.0.gem-old pkg/inq-26.0.0.gem
Original date:      2021-07-29 00:00:00 UTC
Reconstructed date: 2021-07-29 00:00:00 UTC
Reconestructed timestamp: 1627516800
inq 26.0.0 built to pkg/inq-26.0.0.gem.
pkg/inq-26.0.0.gem-old  5cc1697738acf7270777cbfd265db0bbec895cf408e568dd453d079439b863a3
pkg/inq-26.0.0.gem      e0845f8abaefc2ca1958ca0dcd17b8326f16c9c6ee18faa66bc395dc1fa6ce82
~/dev/duckinator/inq$
```

**Comment by duckinator** [comment]
I figured it out, and it means we can't rely on `Gem::Specification#date`:

1. `Gem.source_date_epoch` is an arbitrary timestamp.
2. `Gem::Specification#date` is _just_ a date. The time is always set to 00:00:00 UTC.

This means that there is exactly one second each day where these values are the same. They only match 1/86400 of the time.

**Comment by duckinator** [comment]
Since a `.gem` file is just a renamed tar file, I used the mtime of an arbitrary 

<details>
<summary>Updated version of `gem-rebuild.rb`</summary>

```ruby
```

</details>

I'm more confident in this version, but still can't reproduce a build.

**Comment by duckinator** [comment]
Also: you do need to be using the same RubyGems and Ruby versions. The metadata file contains both the RubyGems version used to build the .gem file and the required Ruby version constraint if any.

So, since the _inq_ gemspec has `spec.required_ruby_version = "~> 2.4"`, I was able to just poke at the metadata, which contains this:

```
required_ruby_version: !ruby/object:Gem::Requirement
  requirements:
  - - "~>"
    - !ruby/object:Gem::Version
      version: '2.4'
required_rubygems_version: !ruby/object:Gem::Requirement
  requirements:
  - - ">="
    - !ruby/object:Gem::Version
      version: '0'
requirements: []
rubygems_version: 3.0.3
```
Fortunately, I didn't need need more precision for the Ruby version -- using the last 2.4.x version was good enough to get this far.

**Comment by duckinator** [comment]
Just edited a some of my comments to clarify a few things.

**Comment by deivid-rodriguez** [comment]
Does the whitespace mismatch issue happen with gems generated with a more recent rubygems?

**Comment by duckinator** [comment]
When I get back to working on this again (hopefully sometime in the next week or two), I'll find something with a more recent release (or make one for one of my projects) that uses the basic `rake build` process and try it.

**Comment by duckinator** [comment]
Yes! A gem with a more modern RubyGems works!

Known-working `gem-rebuild.rb`:

```ruby
#!/usr/bin/env ruby

# Rebuild a gem from source using `rake build`, with SOURCE_DATE_EPOCH
# set to a timestamp derived from the value of the `date` field in the gem
# metadata.

require 'date'
require 'digest'
require 'rubygems/package'

DATE_FORMAT = '%Y-%m-%d %H:%M:%S.%N Z'

def print_sha256(file)
  hash = Digest::SHA256.hexdigest(File.read(file))
  puts "#{hash}\t#{file}"
  hash
end

def get_timestamp(file)
  mtime = nil
  Gem::Package::TarReader.new(File.open(file)) { |tar|
    mtime = tar.seek('metadata.gz') { |f| f.header.mtime }
  }

  mtime
end

def rebuild(original_file, output_file)
  timestamp = get_timestamp(original_file).to_s
  date = Time.at(timestamp.to_i).strftime('%F %T %Z')
  puts "Timestamp: #{timestamp} (#{date})"

  system({"SOURCE_DATE_EPOCH" => timestamp}, "rake build")

  hash1 = print_sha256(original_file)
  hash2 = print_sha256(output_file)

  if hash1 == hash2
    puts "MATCH"
    exit 0
  else
    puts "NO MATCH"
    exit 1
  end
end

if ARGV.length == 2
  rebuild(ARGV[0], ARGV[1])
else
  puts "Usage: $0 ORIGINAL_FILE.gem OUTPUT_FILE.gem"
  puts
  puts "Rebuild a gem from source, with SOURCE_DATE_EPOCH set to a value"
  puts "derived from ORIGINAL_FILE.gem."
  exit 1
end
```

Example usage:

```
~/dev/duckinator/okay$ ../gem-rebuild.rb ~/Downloads/okay-12.0.0.gem pkg/okay-12.0.0.gem
Timestamp: 1628889695 (2021-08-13 17:21:35 EDT)
okay 12.0.0 built to pkg/okay-12.0.0.gem.
70c46fd74688ecd098d6570d58befcfe4dcd4336267c559c030cb0c0d831f7b4        /home/puppy/Downloads/okay-12.0.0.gem
70c46fd74688ecd098d6570d58befcfe4dcd4336267c559c030cb0c0d831f7b4        pkg/okay-12.0.0.gem
MATCH
~/dev/duckinator/okay$
```

**Comment by duckinator** [comment]
As a note: This was built in the same environment. So there's still the problem of figuring out which information needs to be known for reproducing the builds.

But the general approach seems sound, and the script works.

**Comment by deivid-rodriguez** [comment]
Cool, great that you're making progress :muscle:

**Comment by deivid-rodriguez** [comment]
Ok, so I think you would be ready now to transform your script into a `gem rebuild` command and propose a pull request?

**Comment by duckinator** [comment]
@deivid-rodriguez I think so. That'll probably be the next thing I work on.

**Comment by duckinator** [comment]
(Marked this as assigned to myself so I can actually find this issue without searching for like 10 minutes in the future. :sweat_smile: )

**Comment by duckinator** [comment]
Made a draft PR for this over at #4913. I'm pretty happy with how it's come along so far.

**Comment by deivid-rodriguez** [comment]
I think this was fixed by #4913?

**Comment by duckinator** [comment]
Yeah, it was.
